[FEAT] MyPage 프로필 조회 구현

HyeonChae1104 opened the pull request:
## 주요 변경사항 ✔️
기존의 `useGetUser` 훅을 사용하여 구현하였습니다. 

- 헤더의 프로필 아이콘을 누르면 마이페이지로 이동
- 마이페이지에서는 백엔드의 userProfile 정보를 받아와 보여주도록 하였습니다.
- UI 설정
 
## 리뷰어에게... 🙏
코드에 잘못된 부분이 있는지 확인 부탁드립니다.!

## 관련 이슈

closes #64

HyeonChae1104 committed:
feat(Header): 헤더에 프로필 아이콘 클릭 시 마이페이지로 이동 기능 추가

- handleUserIconClick 함수 추가하여 HeaderUserIcon에 onClick으로 전달
로고 클릭 시 메인 페이지로 이동하는 기능
HeaderUserIcon에 onClick 전달

#64

HyeonChae1104 committed:
feat(MyPage):  마이페이지 UI 및 프로필 조회 기능 추가

- - 백엔드에서 사용자 프로필 정보를 받아와 화면에 표시
- 프로필 이미지와 닉네임을 중앙 상단에 배치
- mypage에서 사용자 프로필 이미지를 큰 사이즈로 표시하기 위해 UserProfileImage 컴포넌트에  "lg" 사이즈 옵션을 추가

#64

geongyu09 commented on src/components/features/User/HeaderUserIcon/index.tsx:
```diff
@@ -1,15 +1,35 @@
+};
+
+export const HeaderUserIcon: React.FC<Props> = ({ onClick }) => {
+  const navigate = useNavigate();
```
HeaderUserIcon 컴포넌트의 역할 범위에 대해 고민해보면 좋을 것 같습니다

현재는 유저 아이콘 UI 표시와 클릭 시 리다이렉트 기능을 모두 담당하고 있는데요, 향후 확장성을 위해 책임을 분리하는 것은 어떨까요? 예를 들면:

1. 클릭 동작이 리다이렉트에서 팝업이나 다른 기능으로 변경될 수 있고
2. 사용자 상호작용도 클릭 외에 호버나 다른 방식으로 확장될 수 있습니다

이런 변경사항에 유연하게 대응하려면, HeaderUserIcon은 UI 표시에 집중하고 상호작용 로직은 상위 컴포넌트로 위임하는 구조가 더 좋지 않을까 생각합니다.

HyeonChae1104 replied:
넵! 말씀대로 HeaderUserIcon은 UI에 집중하는 것이 좋을 것 같아 수정해보았습니다!

geongyu09 commented on src/components/features/User/HeaderUserIcon/index.tsx:
```diff
@@ -1,15 +1,35 @@
       : profileImage.profile_image_url;
 
-  return <UserProfileImage profileImageUrl={profile} />;
+  const handleClick = () => {
```
HeaderUserIcon는 다른 곳에서 사용중입니다. 그리고 앞으로 서비스가 커지게 되면 다른 곳에서도 사용될 수 있는 컴포넌트입니다. 
그렇기에 최대한 사이드 이팩트를 줄여야 하고 다른 동작을 하지 않도록 해야합니다. 

간단히 예시를 들어보면,
현재 이 컴포넌트는 그룹 관리 페이지에서 참여자를 보여주기 위해 사용중입니다. (실제로 사용중입니다)
그런데 클릭시 마이페이지로 이동하도록 만들어 버리면 내가 아닌 다른 참여자를 클릭해도 내 마이페이지로 이동하게 됩니다.

반대로 미리 navigate가 있는 컴포넌트였다면, 저는 다른 사용자를 보여줘야하는 컴포넌트를 만들때 이 컴포넌트를 사용할 수 없겠죠? 원하는 동작을 하지 않으니까요

Da-Yeon-Kim replied:
엇 그룹 관리 페이지에서 `HeaderUserIcon`을 사용하고 계시나요? 이 컴포넌트는 헤더에서만 사용해서 `HeaderUserIcon`이란 이름을 가지고 있는 줄 알았는데요!

그룹 페이지는 common의 `Participant`를 사용하고 있는 것 같아 보이는데, 음 제가 잘못 안 건가요...?

HyeonChae1104 replied:
반영하였습니다!

geongyu09 commented on src/pages/MyPage/index.tsx:
```diff
@@ -1,3 +1,59 @@
+import { UserProfileImage } from '@/components/common/User/UserProfileImage';
+
+export const MyPage: React.FC = () => {
+  const { data: profileData, status } = useGetUserProfile();
```
page 컴포넌트에서는 되도록이면 쿼리문을 지양하는게 좋아보입니다. (지금것 저희 프로젝트가 그래왔었고요)
presentation-container 패턴으로 작성할 때는 모든 데이터들을 상위 컴포넌트에 두고, 하위 컴포넌트는 이를 내려 받아서 보여주기만 하는 방식으로 짜게 됩니다. 
하지만 이 패턴은 커스텀 훅이 나오면서 안티패턴 느낌이 되어버렸습니다. (안티 패턴으로 낙인찍힌건 아니지만, 훨씬 좋은 패턴을 두고 복잡하게 사용할 이유는 없기 때문에 여론이 부정적입니다)

요즘은 커스텀 훅으로(useGetUserProfile도 커스텀훅 패턴입니다) 필요한 곳에 가져다 사용하는 방식으로 개발합니다.

지금은 페이지 내부에 하나의 요청만 필요할지 모르지만, 한 페이지에 정말 많은 기능이 들어갈 수 있습니다. 
이 경우 책임별로 section을 나누어서 개발하는것이 일반적입니다. 
그리고 해당 section마다 필요한 쿼리문을 가져다 사용하는 방식으로 개발하게 됩니다.

다른 페이지 모두 동일한 방식으로 구성되어있으니 참고하면 좋을 것 같습니다

HyeonChae1104 replied:
자세하게 설명해주셔서 정말 감사합니다!
말씀하신 방식으로 수정해보았습니다!

geongyu09 reviewed (changes requested):
수고 많으셨습니다!
전체적으로 코드가 깔끔하네요
다만코드 퀄리티 관련 사항과 공통 컴포넌트 수정으로 인해서 문제가 야기될 수 있어서 수정되어야 할 것 같습니다

Da-Yeon-Kim commented on src/components/features/Layout/Header/index.tsx:
```diff
@@ -16,10 +16,14 @@ export const Header: React.FC<Props> = ({ height }) => {
   };
 
+  const handleUserIconClick = () => {
+    router('/mypage');
```
기존에 만들어둔 `RouterPath`를 사용하시는 쪽을 권장 드립니다! `@/routes` 폴더 내에 있답니다ㅎㅎ 변경하신다면, 그 김에 위에 `/`도 바꿔주시면 더 좋을 것 같아요,, ㅎㅎ

HyeonChae1104 replied:
네! 수정하였습니다!

Da-Yeon-Kim reviewed (approved):
고생하셨어요~😃
깔끔하게 잘 보이네요! 사이즈 하나 더 추가하신 것도 좋아요!

건규 씨가 언급해주신 대로 계층 구조에 맞게 훅 사용 위치만 좀 조정해주시면 좋을 것 같아요~

HyeonChae1104 committed:
refactor(HeaderUserIcon): UI와 클릭로직의 책임 분리

- HeaderUserIcon은 UI를 담당,
- Header에서 클릭 시 마이페이지로 이동하는 것을 수행하도록 분리하였습니다.

#64

HyeonChae1104 committed:
refactor(MyPage): presentation-container 패턴 적용

- useGetUserProfile 훅을 사용해 데이터를 가져오고, MyPage는 props로 전달받아 렌더링만 담당
- MyPage 컴포넌트를 상위 MyPageContainer에서 데이터 로직을 처리하도록 리팩토링하였습니다.

#64

HyeonChae1104 committed:
refactor(HeadUserIcon): HeaderUserIcon 컴포넌트 재사용성 개선

- HeaderUserIcon에서 navigate 기능을 제거하고, 프로필 이미지를 렌더링하는 역할에 집중,
- Header 컴포넌트에서 RouterPath 객체 사용

#64

HyeonChae1104 commented:
리뷰 해주신 부분들은 전부 수정해보았습니다!

geongyu09 commented on src/components/features/User/HeaderUserIcon/index.tsx:
```diff
@@ -1,15 +1,20 @@
-  const { data: profileImage, status } = useGetUserProfile();
+type Props = {
+  profileImageUrl?: string;
+  onClick?: () => void;
```
onClick도 사실 없으면 좋을 것 같긴 합니다. 
onClick이 아니라 다른 이벤트 리스터가 추가된다면, 이벤트 리스너 뿐만 아니라 다른 기능을 추가하게 된다면 계속해서 props가 추가될텐데, 분명히 유지보수적으로 문제가 있을거에요

아니면 차라리 다른 공통 컴포넌트처럼 HTML속성 전체를 상속받은 방법도 존재하는데, 유저 아이콘을 보여주는 컴포넌트에 div의 모든 속성을 가지고 있을 필요가 있을까? 싶기도 합니다

onClick도 없애고, 상위에서 사용할 때 Link태그를 사용하면 되지 않을까? 싶습니다

물론 현 상황에서는 문제가 생기지 않기에 추후에 변경해도 좋아보입니다

geongyu09 commented on src/components/features/MyPageContainer/index.tsx:
```diff
@@ -0,0 +1,20 @@
+
+import { useGetUserProfile } from '@/api/hooks/useGetUser';
+
+export const MyPageContainer: React.FC = () => {
```
container보단 section과 같은 이름이면 좋을 것 같습니다!
container는 다른 컨텐츠들을 담는 컴포넌트에 주로 이름을 붙이는 느낌이라 해당 컴포넌트와는 조금 안맞는 느낌이 있는 것 같습니다..!

geongyu09 commented on src/routes/index.tsx:
```diff
@@ -40,7 +40,7 @@ const router = createBrowserRouter([
       {
         path: RouterPath.mypage,
-        element: <MyPage />,
+        element: <MyPageContainer />, // MyPage 대신 MyPageContainer 사용
```
route에서 page가 아닌 컴포넌트를 부르고 있네요. 
의도하신 부분이어도 문제가 있습니다. page컴포넌트가 최상위가 되어야합니다

geongyu09 reviewed (changes requested):
지금 route/index.ts에서 page컴포넌트가 아닌 다른 컴포넌트를 부르고 있네요. 
해당 부분 수정이 필요합니다

HyeonChae1104 committed:
refactor(HeaderUserIcon): onClick 제거

- onClick 제거
- Header에서 Link태그를 사용하도록 변경하였습니다.

#64

HyeonChae1104 committed:
refactor(MyPage): page 컴포넌트를 부르도록 routes 수정

- Container에서 Section으로 컴포넌트 이름 변경
- page 컴포넌트가 최상위가 되도록 수정하였습니다.

#64

HyeonChae1104 commented:
- Container에서 Section으로 컴포넌트 이름 변경
- page 컴포넌트가 최상위가 되도록 라우트 및 코드 수정
- HederUserIcon은 기존의 건규님이 작성하신 코드로 변경하였습니다.(onClick 제거)
- `Header/index.tsx`에서 Link를 이용하여 사용하도록 하였습니다!

geongyu09 commented on src/components/features/Layout/Header/index.tsx:
```diff
@@ -1,25 +1,30 @@
 
 export const Header: React.FC<Props> = ({ height }) => {
-  const router = useNavigate();
+  const navigate = useNavigate(); // useNavigate 훅을 사용하여 navigate를 선언
```
이거 수정했어야 했는데.. 감사합니다

geongyu09 commented on src/components/features/Layout/Header/index.tsx:
```diff
@@ -1,25 +1,30 @@
 
   const handleLogoClick = () => {
-    router('/');
+    navigate(RouterPath.home);
```
👍👍👍👍

geongyu09 commented on src/components/features/Layout/Header/index.tsx:
```diff
@@ -1,25 +1,30 @@
-      <HeaderUserIcon />
+      <Link to={RouterPath.mypage}>
+        <HeaderUserIcon />
+      </Link>
```
👍👍👍👍👍👍👍👍

geongyu09 commented on src/components/features/MyPageSection/index.tsx:
```diff
@@ -0,0 +1,18 @@
+    return <div>프로필 정보를 불러오는 데 실패했습니다.</div>;
+  }
+
+  return <MyPage profileData={profileData} />;
```
section 컴포넌트에서 페이지 컴포넌트를 호출하는 이유가 있나요?

geongyu09 commented on src/components/features/MyPageSection/index.tsx:
```diff
@@ -0,0 +1,18 @@
+
+import { useGetUserProfile } from '@/api/hooks/useGetUser';
+
+export const MyPageSection: React.FC = () => {
```
현재 사용되지 않는 컴포넌트인 것 같은데 제거해도 괜찮아보이네요!

geongyu09 commented on src/pages/MyPage/index.tsx:
```diff
@@ -1,3 +1,40 @@
+  nickname: string;
+}
+
+interface MyPageProps {
```
페이지 컴포넌트에서 props를 받으면, 이는 어디에 사용되나요? 
페이지 컴포넌트에서 props를 주려면 route/index.tsx 에서 라우팅 로직에서 props를 주어야 하는데, 그러면 라우팅 로직에서 profileData를 주기 위해서 또 서버에서 요청을 해야 하고, 그러면 라우팅만 하는 코드에 또 다른 책임이 생기게 됩니다

props를 받는 코드이지만, 실제적으로는 props를 전달하고 있지 않네요.

geongyu09 commented on src/pages/MyPage/index.tsx:
```diff
@@ -1,3 +1,40 @@
+
+export const MyPage: React.FC<MyPageProps> = ({ profileData }) => (
+  <ProfileContainer>
+    <UserProfileImage profileImageUrl={profileData?.profile_image_url || '/images/default_profile_image.png'} size="lg" />
```
'/images/default_profile_image.png' 파일이 존재하지 않습니다. 
추가적으로 해당 유저의 url이 없으면 기본 이미지를 보여주도록 하는 로직은 이미 UserProfileImage에 구현되어있습니다

geongyu09 reviewed (changes requested):
다시한번 수정이 필요해보이네요.
올바르게 사용자의 프로필을 받아오고 있지 않습니다
props로 받은 이미지를 보여주고 있는데 props를 전달하고 있지 않고, page 컴포넌트에서 props를 전달하고 있습니다.

코드의 퀄리티적인 부분은 수정하지 않아도 괜찮지만, 돌아가는 코드가 되어야 합니다

Da-Yeon-Kim reviewed (commented):
기존에 `MyPageSection` 컴포넌트에서 `useGetUserProfile`을 사용해서 백엔드로부터 유저의 프로필을 가져오고 있었는데,, 이 컴포넌트를 사용하지 않고 있네요!

사용하지 않는 `MyPageSection`은 우선 삭제하시고, `components/features/MyPage` 폴더로 바꾸신 다음에 그 안에 마이페이지 관련된 컴포넌트들을 분리해 넣으시는 게 좋을 것 같습니다. 다른 페이지를 참고해보세요😃

마이페이지를 만약 제가 만든다면, Profile, Preferences, nonPreferences 이렇게 세 가지 섹션으로 나눌 것 같습니다! 유저의 정보를 가져오는 훅(`useGetUser`) 같은 경우에는 Profile 섹션에 두고 사용할 것 같아요. 나머지 선호/비선호 음식 조회 API도 마찬가지로 각 섹션 하위 컴포넌트에서 할 것 같네요.
(한 번 참고하시라고 if의 상황을 생각해보았습니다ㅎㅎ)

또한, 페이지는 보통 최상위 컴포넌트이기 때문에 props를 받지 않고, 다른 컴포넌트에서 사용하지도 않습니다! 하위 컴포넌트를 모아서 보여주는 역할을 하고, 여러 하위 컴포넌트에서 사용되는 정보가 있는 경우 페이지에서 받아와 나눠주기도 합니다~

HyeonChae1104 committed:
refactor(MyPage): presentation-container 패턴 적용 수정

- useGetUserProfile 사용하여 사용자 데이터 가져옵니다.
- UserProfileImage의 로직 사용
- ProfileSection 컴포넌트에서 데이터 관련 부분을 관리하고 `MyPage/index.tsx`에서는 해당 섹션을 불러와 렌더링합니다.

#64

HyeonChae1104 commented:
제가 너무 부족하네요,, 두분이 항상 꼼꼼하게 리뷰해주셔서 늘 감사하게 생각하고 있습니다ㅜ 

- `useGetUserProfile` 훅 사용하여 유저 프로필 데이터 받도록 함.
- `UserProfileImage` 에서 구현된 로직 사용 (기본 이미지를 보여주도록 하는 로직 등은 제 코드에서 삭제하였습니다.!)
- 파일 구조 변경 : 다연님이 말씀해주신대로 `components/MyPage/ProfileSection` 으로 폴더를 구성하였고, 이후 선호 비선호 섹션도 해당 부분에서 분리하여 구현하는 것으로 수정하겠습니다. 

확인 부탁드립니다..

geongyu09 commented on src/components/features/MyPage/ProfileSection/index.tsx:
```diff
@@ -0,0 +1,52 @@
+
+  return (
+    <ProfileContainer>
+      <UserProfileImage profileImageUrl={profileData.profile_image_url} size="lg" />
```
지금도 정말 훌륭한 코드이지만, 
profileData.profile_image_url, profileData.nickname 처럼 profileData. 이 반복되므로 
const {profile_image_url, nickname} = profileData; 문장을 추가해서 profile_image_url, nickname 형태로 사용해도 좋아보이네요!

수정할 필요까지 없어보입니다

geongyu09 reviewed (approved):
완벽한 코드가 되었네요!
수정하느라 고생 많으셨습니다...!

Da-Yeon-Kim commented on src/pages/MyPage/index.tsx:
```diff
@@ -1,3 +1,12 @@
+
+export const MyPage: React.FC = () => {
+  return (
+    <div>
```
여기에는 `div` 대신에 건규 씨가 만들어주신 `@/components/common/layouts/Container`를 사용하면 더 좋을 것 같긴 하네요! (대신 이 경우에는 아마 ProfileSection 스타일링도 살짝 수정해야 할 거예요) 아니면 따로 커스텀 하셔도 되고요~

HyeonChae1104 replied:
이 부분은 제일 마지막에,, 반영해보도록 하겠습니다! 
알려주셔서 감사합니다!! 👍👍

Da-Yeon-Kim commented on src/pages/MyPage/index.tsx:
```diff
@@ -1,3 +1,12 @@
-export const MyPage = () => {
-  return <div>MyPage</div>;
+import React from 'react';
+import { ProfileSection }from 'src/components/features/MyPage/ProfileSection';
```
여기도 절대 경로를 사용해서 `@/components/features/MyPage/ProfileSection`으로 작성하시면 더 좋겠네요!

HyeonChae1104 replied:
넵! 반영하였습니다!!

Da-Yeon-Kim reviewed (approved):
엄청 좋은 코드가 되었네요!
PR 리뷰를 수용하려고 하시는 태도 항상 굳입니다👍

리뷰한 사항은 꼭 변경해야 하는 부분은 아닙니다!
나중에 음식 부분 개발하시면서 수정해도 되고요~

HyeonChae1104 committed:
refactor(MyPage): 절대 경로로 import 경로 수정

#64

HyeonChae1104 commented:
절대 경로로 import 경로 수정하였습니다.! 감사합니다
다시 리뷰는 받지 않고.! 바로 머지하도록 하겠습니다!
